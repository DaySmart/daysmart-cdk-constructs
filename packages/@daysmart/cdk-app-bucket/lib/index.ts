import cdk = require('@aws-cdk/core');
import s3 = require('@aws-cdk/aws-s3');

export class AppBucket extends cdk.Construct {
    constructor(scope: cdk.Construct, id: string, props: any) {
        super(scope, id);

        new s3.Bucket(this, 'Bucket', {
            bucketName: `${props.stage}-${props.appName.toLowerCase()}-website-bucket`,
            publicReadAccess: true,
            versioned: true,
        });
    }
}