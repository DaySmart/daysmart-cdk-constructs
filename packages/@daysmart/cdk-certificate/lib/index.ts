import * as cdk from '@aws-cdk/core';
import acm = require('@aws-cdk/aws-certificatemanager');
import route53 = require('@aws-cdk/aws-route53');

export interface CdkCertificateProps {
    domainName: string;
    stage: string;
    project: string;
}

export class CdkCertificate extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkCertificateProps) {
    super(scope, id);

    const hostedZone = route53.HostedZone.fromLookup(this, 'HostedZone', {
        domainName: props.domainName
    });

    new acm.Certificate(this, 'Certificate', {
        domainName: props.domainName,
        subjectAlternativeNames: [
            `${props.stage}.${props.project}.${props.domainName}`,
            `*.${props.stage}.${props.project}.${props.domainName}`
        ],
        validation: acm.CertificateValidation.fromDns(hostedZone)
    });
  }
}
