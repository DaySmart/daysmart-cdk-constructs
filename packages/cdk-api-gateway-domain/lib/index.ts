import * as cdk from '@aws-cdk/core';
import * as apigw from '@aws-cdk/aws-apigateway';
import * as acm from '@aws-cdk/aws-certificatemanager';
import * as route53 from '@aws-cdk/aws-route53';

export interface CdkApiGatewayDomainProps {
  companyDomainName: string;
  companyHostedZoneId: string;
  project: string;
  baseEnv: string;
  dynamicEnv?: string;
  certificateArn: string;
  restApiId: string;
  basePath: string;
}

export class CdkApiGatewayDomain extends cdk.Construct {

  constructor(scope: cdk.Construct, id: string, props: CdkApiGatewayDomainProps) {
    super(scope, id);

    const api = apigw.RestApi.fromRestApiId(this, "Rest API", `${props.restApiId}`);

    let customDomain: apigw.DomainName;

    if (props.dynamicEnv) {
      customDomain = new apigw.DomainName(this, 'Custom Domain', {
        domainName: `${props.dynamicEnv}-api.${props.baseEnv}.${props.project}.${props.companyDomainName}`,
        certificate: acm.Certificate.fromCertificateArn(this, "Certificate", `${props.certificateArn}`),
        endpointType: apigw.EndpointType.EDGE,
        securityPolicy: apigw.SecurityPolicy.TLS_1_0
      });
    } else {
      customDomain = new apigw.DomainName(this, 'Custom Domain', {
        domainName: (props.baseEnv == "prod") ? `api.${props.project}.${props.companyDomainName}` : `api.${props.baseEnv}.${props.project}.${props.companyDomainName}`,
        certificate: acm.Certificate.fromCertificateArn(this, "Certificate", `${props.certificateArn}`),
        endpointType: apigw.EndpointType.EDGE,
        securityPolicy: apigw.SecurityPolicy.TLS_1_0
      });
    }

    const cloudformationBasePathMapping = new apigw.CfnBasePathMapping(this, "CloudformationBasePathMapping", {
      basePath: `${props.basePath}`,
      domainName: customDomain.domainName,
      restApiId: `${props.restApiId}`,
      stage: `${props.baseEnv}` 
    });

    const cloudformationRoute53Change = new route53.CfnRecordSet(this, "CloudformationAliasRecordRoute53", {
      aliasTarget: {
        dnsName: customDomain.domainNameAliasDomainName,
        evaluateTargetHealth: false,
        hostedZoneId: "Z2FDTNDATAQYW2"
      },
      comment: `CloudFront distribution for ${customDomain.domainName}`,
      hostedZoneId: props.companyHostedZoneId,
      name: customDomain.domainName,
      type: "A"
    });
  }
}
