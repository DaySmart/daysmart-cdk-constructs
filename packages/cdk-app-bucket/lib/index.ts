import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_cloudfront as cloudfront } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';

export interface AppBucketProps {
    stage: string;
    appName: string;
    dynamicEnvName: string;
    projectName: string;
    sharedServicesAccountId?: string;
}
export class AppBucket extends Construct {
    constructor(scope: Construct, id: string, props: AppBucketProps) {
        super(scope, id);

        const bucket = new s3.Bucket(this, 'Bucket', {
            bucketName: `${props.dynamicEnvName}-${props.appName}.${props.stage}.${props.projectName}`,
            versioned: true,
            objectOwnership: s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
            blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL
        });

        const originAccessIdentity = new cloudfront.OriginAccessIdentity(this, 'OriginAcessIdentity', {
            comment: `OriginAccessIdentity for ${bucket.bucketName}.`
        });

        bucket.grantRead(originAccessIdentity)

        if(props.sharedServicesAccountId) {
            bucket.grantReadWrite(new iam.AccountPrincipal(props.sharedServicesAccountId));
        }

        let output = new cdk.CfnOutput(this, "AppBucket", {
            value: bucket.bucketName
        });

        let output2 = new cdk.CfnOutput(this, "OriginAccessIdentity", {
            value: originAccessIdentity.originAccessIdentityName
        });

        output.overrideLogicalId("AppBucket");
        output2.overrideLogicalId("OriginAccessIdentity");
    }
}