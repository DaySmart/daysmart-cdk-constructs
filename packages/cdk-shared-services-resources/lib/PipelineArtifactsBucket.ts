import * as cdk from 'aws-cdk-lib'
import { Construct } from 'constructs'
import { aws_s3 as s3 } from 'aws-cdk-lib';
import { aws_iam as iam } from 'aws-cdk-lib';
import { aws_kms as kms } from 'aws-cdk-lib';
import * as assert from 'aws-cdk-lib/assertions'


/**
 * Properties for shared S3 bucket for pipeline artifacts
 */
export interface PipelineArtifactsBucketProps {

    /**
     * The name of the application or group of applications the bucket will be used for
     * 
     */
    applicationGroup: string;

    /**
     * The 2 letter product group code
     * 
     */
    productGroup: string;

    /**
     * The Arn for the shared KMS key used for bucket encryption. Use SharedKMSKey construct
     * 
     */
    kmsKeyId: string;

    /**
     * List of AWS account ids that require read permissions to the bucket
     * 
     */
    readAccountIds?: string[];
}

/**
 * A shared S3 bucket for storing artifacts from CodePipeline
 */
export class PipelineArtifactsBucket extends Construct {
    readonly artifactBucket: s3.IBucket;

    constructor(scope: Construct, id: string, props: PipelineArtifactsBucketProps) {
        super(scope, id);

        this.artifactBucket = new s3.Bucket(this, 'ArtifactBucket', {
            bucketName: `${props.productGroup}-${props.applicationGroup}-pipeline-artifacts-${cdk.Stack.of(this).region}`,
            versioned: true,
            encryption: s3.BucketEncryption.KMS,
            accessControl: s3.BucketAccessControl.PRIVATE,
            encryptionKey: kms.Key.fromKeyArn(this, 'SharedKey', props.kmsKeyId)
        });

        if(props.readAccountIds && props.readAccountIds.length > 0) {
            props.readAccountIds.forEach(accountId => {
                this.artifactBucket.grantRead(new iam.AccountPrincipal(accountId));
            });
        }

        const bucketArn = new cdk.CfnOutput(this, 'BucketArn', {
            value: this.artifactBucket.bucketArn
        });
        bucketArn.overrideLogicalId('BucketArn');
    }
}