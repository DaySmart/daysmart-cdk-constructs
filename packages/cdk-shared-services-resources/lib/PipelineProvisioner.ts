import * as cdk from 'aws-cdk-lib/core';
import * as codebuild from 'aws-cdk-lib/aws-codebuild';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

/**
 * Properties for a CodeBuild project to provision CodePipeline pipelines for all branches in a given repo
 * 
 */
export interface PipelineProvisionerProps {
    /**
     * Owner of the repository
     */
    owner: string;

    /**
     * Name of the repository
     */
    repo: string;
}

/**
 * CodeBuild project with the necessary permissions to provision a CodePipeline for each branch
 */
export class PipelineProvisioner extends Construct {
    readonly buildProject: codebuild.IProject;

    constructor(scope: Construct, id: string, props: PipelineProvisionerProps) {
        super(scope, id);

        const role = new iam.Role(this, 'Role', {
            roleName: `pipeline-provisioner-role-${props.repo}`,
            description: 'Enables the pipeline provisioner CodeBuild projects to provision new CodePipelines',
            assumedBy: new iam.ServicePrincipal('codebuild.amazonaws.com')
        });

        role.attachInlinePolicy(new iam.Policy(this, 'Policy', {
            statements: [
                new iam.PolicyStatement({
                    effect: iam.Effect.ALLOW,
                    actions: [
                        'cloudformation:*',
                        'codepipeline:*',
                        'codebuild:*',
                        'codedeploy:*',
                        'codestar:*',
                        'codestar-connections:*',
                        's3:List*',
                        's3:Get*',
                        's3:Describe',
                        'iam:Attach*',
                        'iam:Create*',
                        'iam:Get*',
                        'iam:List*',
                        'iam:PassRole',
                        'iam:Put*',
                        'lambda:*'
                    ],
                    resources: ['*']
                })
            ]
        }));

        this.buildProject = new codebuild.Project(this, 'Project', {
            projectName: `${props.repo}-pipeline-provisioner`,
            environment: {
                buildImage: codebuild.LinuxBuildImage.STANDARD_5_0,
                computeType: codebuild.ComputeType.SMALL
            },
            buildSpec: codebuild.BuildSpec.fromObject({
                version: '0.2',
                phases: {
                    install: {
                        'runtime-versions': {
                            nodejs: 14
                        },
                        commands: [
                            'cd infrastructure/pipeline',
                            'npm i'
                        ]
                    },
                    build: {
                        commands: [
                            'npx cdk deploy --require-approval never'
                        ]
                    }
                }
            }),
            concurrentBuildLimit: 1,
            source: codebuild.Source.bitBucket({
                owner: props.owner,
                repo: props.repo,
                webhook: true,
                webhookFilters: [
                    codebuild.FilterGroup.inEventOf(
                        codebuild.EventAction.PUSH
                    )
                ]
            }),
            role: role,
            badge: false
        })
    }
}