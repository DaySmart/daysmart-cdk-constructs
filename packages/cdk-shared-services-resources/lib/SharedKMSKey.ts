import * as cdk from 'aws-cdk-lib/core';
import * as kms from 'aws-cdk-lib/aws-kms';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';

/**
 * Properties for the shared KMS key
 */
export interface SharedKMSKeyProps {

    /**
     * The AWS organization id to share the key with
     */
    organizationId: string;

    /**
     * The application or application group the key is for
     */
    applicationGroup: string;

    /**
     * The two letter product group code
     */
    productGroup: string;
}

/**
 * KMS key designed to be shared accross an organization
 */
export class SharedKMSKey extends Construct {
    readonly key: kms.IKey;

    constructor(scope: Construct, id: string, props: SharedKMSKeyProps) {
        super(scope, id);

        this.key = new kms.Key(this, 'Key', {
            description: "KMS Key for Encryption of Resources",
            enabled: true
        });

        this.key.grantDecrypt(new iam.OrganizationPrincipal(props.organizationId));

        const alias = this.key.addAlias(`alias/${props.productGroup}-${props.applicationGroup}-shared-services-${cdk.Stack.of(this).region}`);

        const keyArn = new cdk.CfnOutput(this, 'KeyArn', {
            value: this.key.keyArn
        });
        keyArn.overrideLogicalId('KeyArn');

        const keyAlias = new cdk.CfnOutput(this, 'KeyAlias', {
            value: alias.aliasName
        });
        keyAlias.overrideLogicalId('KeyAlias');
    }
}