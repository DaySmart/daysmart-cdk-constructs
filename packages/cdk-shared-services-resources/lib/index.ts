export { SharedKMSKey, SharedKMSKeyProps } from './SharedKMSKey';
export { PipelineArtifactsBucket, PipelineArtifactsBucketProps } from './PipelineArtifactsBucket';
export { PipelineProvisioner, PipelineProvisionerProps } from './PipelineProvisioner'; 
