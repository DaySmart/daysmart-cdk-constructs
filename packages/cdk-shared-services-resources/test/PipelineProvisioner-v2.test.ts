import * as cdk from 'aws-cdk-lib';
import { Template } from "aws-cdk-lib/assertions"
import { PipelineProvisioner} from '../lib/index'

test('App Cloudfront', () => {
    const stack = new cdk.Stack(undefined, 'stack', {
        env: {
            account: '123456',
            region: 'us-east-1'
        }
    });
    new PipelineProvisioner(stack, 'AppCloudfront', {
        owner: 'tester',
        repo: 'cdk'
    });

    const template = Template.fromStack(stack);
    console.log(JSON.stringify(template, null, 2));

    expect(Template.fromStack(stack)).toMatchSnapshot();
})