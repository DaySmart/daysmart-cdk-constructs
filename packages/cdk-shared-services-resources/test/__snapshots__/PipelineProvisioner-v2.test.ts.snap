// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Cloudfront 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "AppCloudfrontPolicy6835BCDC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "cloudformation:*",
                "codepipeline:*",
                "codebuild:*",
                "codedeploy:*",
                "codestar:*",
                "codestar-connections:*",
                "s3:List*",
                "s3:Get*",
                "s3:Describe",
                "iam:Attach*",
                "iam:Create*",
                "iam:Get*",
                "iam:List*",
                "iam:PassRole",
                "iam:Put*",
                "lambda:*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AppCloudfrontPolicy6835BCDC",
        "Roles": Array [
          Object {
            "Ref": "AppCloudfrontRoleB0CF141D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "AppCloudfrontProjectEE854E63": Object {
      "Properties": Object {
        "Artifacts": Object {
          "Type": "NO_ARTIFACTS",
        },
        "BadgeEnabled": false,
        "Cache": Object {
          "Type": "NO_CACHE",
        },
        "ConcurrentBuildLimit": 1,
        "EncryptionKey": "alias/aws/s3",
        "Environment": Object {
          "ComputeType": "BUILD_GENERAL1_SMALL",
          "Image": "aws/codebuild/standard:5.0",
          "ImagePullCredentialsType": "CODEBUILD",
          "PrivilegedMode": false,
          "Type": "LINUX_CONTAINER",
        },
        "Name": "cdk-pipeline-provisioner",
        "ServiceRole": Object {
          "Fn::GetAtt": Array [
            "AppCloudfrontRoleB0CF141D",
            "Arn",
          ],
        },
        "Source": Object {
          "BuildSpec": "{
  \\"version\\": \\"0.2\\",
  \\"phases\\": {
    \\"install\\": {
      \\"runtime-versions\\": {
        \\"nodejs\\": 14
      },
      \\"commands\\": [
        \\"cd infrastructure/pipeline\\",
        \\"npm i\\"
      ]
    },
    \\"build\\": {
      \\"commands\\": [
        \\"npx cdk deploy --require-approval never\\"
      ]
    }
  }
}",
          "Location": "https://bitbucket.org/tester/cdk.git",
          "ReportBuildStatus": true,
          "Type": "BITBUCKET",
        },
        "Triggers": Object {
          "FilterGroups": Array [
            Array [
              Object {
                "Pattern": "PUSH",
                "Type": "EVENT",
              },
            ],
          ],
          "Webhook": true,
        },
      },
      "Type": "AWS::CodeBuild::Project",
    },
    "AppCloudfrontRoleB0CF141D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "codebuild.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Enables the pipeline provisioner CodeBuild projects to provision new CodePipelines",
        "RoleName": "pipeline-provisioner-role-cdk",
      },
      "Type": "AWS::IAM::Role",
    },
    "AppCloudfrontRoleDefaultPolicy14789CC7": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "AppCloudfrontProjectEE854E63",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":logs:us-east-1:123456:log-group:/aws/codebuild/",
                      Object {
                        "Ref": "AppCloudfrontProjectEE854E63",
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "codebuild:CreateReportGroup",
                "codebuild:CreateReport",
                "codebuild:UpdateReport",
                "codebuild:BatchPutTestCases",
                "codebuild:BatchPutCodeCoverages",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":codebuild:us-east-1:123456:report-group/",
                    Object {
                      "Ref": "AppCloudfrontProjectEE854E63",
                    },
                    "-*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "AppCloudfrontRoleDefaultPolicy14789CC7",
        "Roles": Array [
          Object {
            "Ref": "AppCloudfrontRoleB0CF141D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
